src/platform/plugins/shared/discover/public/application/main/state_management/discover_app_state_container.ts

```diff
@@ -217,9 +218,46 @@ export const getDiscoverAppStateContainer = ({
 
       previousState = appStateContainer.getState();
 
-      // When updating to an ES|QL query, sync the data source
+      // special handling for ES|QL queries
       if (isOfAggregateQueryType(value.query)) {
+        // When updating to an ES|QL query, sync the data source
         value = { ...value, dataSource: createEsqlDataSource() };
+
+        if (services.discoverFeatureFlags.getCascadeLayoutEnabled()) {
+          const currentTabState = selectTab(internalState.getState(), tabId);
+
+          // compute and set cascade groupings when state updates happen
+          internalState.dispatch(
+            injectCurrentTab(internalStateActions.setCascadeUiState)({
+              cascadeUiState: {
+                availableCascadeGroups: getESQLStatsQueryMeta(
+                  (value!.query as AggregateQuery).esql
+                ).groupByFields.map((group) => group.field),
+                get selectedCascadeGroups() {
+                  if (
+                    !currentTabState.uiState.cascadedDocuments ||
+                    (currentTabState.uiState.cascadedDocuments &&
+                      // if the proposed available groups is different in length or contains a value the existing one doesn't have, we want to reset by defaulting to the first group
+                      (this.availableCascadeGroups.length !==
+                        currentTabState.uiState.cascadedDocuments.availableCascadeGroups.length ||
+                        this.availableCascadeGroups.some(
+                          (group) =>
+                            (
+                              currentTabState.uiState.cascadedDocuments?.availableCascadeGroups ??
+                              []
+                            ).indexOf(group) < 0
+                        )))
+                  ) {
+                    return [this.availableCascadeGroups[0]].filter(Boolean);
+                  }
+
+                  // return existing selection since we've asserted that there's been no change to the available groups default
+                  return currentTabState.uiState.cascadedDocuments!.selectedCascadeGroups;
+                },
+              },
+            })
+          );
+        }
```

Comment:
computes the available cascade groups on every change to global state if query type is ES|QL, with consideration for previous selected values especially if the query itself didn't change